### PR TITLE
toaster and toast components

### LINKED ISSUE #1100: Toaster
Add Toaster (see https://evergreen.segment.com/components/toaster as inspiration)

Screenshot 2022-01-12 at 08 54 33

- [x] export { toaster } from components/Toaster
- [x] add methods toaster.info() toaster.success() toaster.warning() toaster.error() 
- [x] signature of methods above:
```typescript
toaster.info(message:string, props?:ToasterProps

type ToasterProps = {
  description?:string (default null)
  duration?:number (in milliseconds) (default 2000)
}
```
- [x] use toaster in axios error middleware with message:
'Error while processing the request' and description the message from the exception

### PR DESCRIPTION
**Pull request by rfontanarosa**
close #1100

Schermata 2022-02-09 alle 12 05 42

### CODE REVIEW

**Comment by sorja** [comment]
Use existing cross (x) styling from country selector

image

File: client/store/ui/notification/stateType.ts [hunk 1]
```diff
@@ -0,0 +1,10 @@
+export interface NotificationMessage {
+  duration?: number
+  id: string
```
Comment (sorja) [inline client/store/ui/notification/stateType.ts:3]: how do you manage id when creating new toaster?
ex. on ajax error?

would it be simpler to use array index?

File: client/store/ui/notification/hooks/index.ts [hunk 2]
```diff
@@ -0,0 +1,4 @@
+import { useAppSelector } from '@client/store'
+import { NotificationState } from '../stateType'
+
+export const useNotification = (): NotificationState | undefined => useAppSelector((state) => state.ui.notification)
```
Comment (sorja) [inline client/store/ui/notification/hooks/index.ts:4]: notification can't be undefined (the initialState is set to empty array)

File: client/components/Toaster/Toaster.scss [hunk 3]
```diff
@@ -0,0 +1,60 @@
+@import 'client/style/partials';
+
+.toaster-container {
+  max-width: 560px;
```
Comment (minotogna) [inline client/components/Toaster/Toaster.scss:4]: on mobile 560 px?

File: client/components/Toaster/Toaster.scss [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+  margin: 0 auto;
+  top: 0;
+  left: 0;
+  right: 0;
+  position: fixed;
+  z-index: 30;
```
Comment (minotogna) [inline client/components/Toaster/Toaster.scss:10]: use z-index constants

File: client/components/Toaster/Toaster.scss [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+  // pointer-events: none;
```
Comment (minotogna) [inline client/components/Toaster/Toaster.scss:11]: remove comments

File: client/store/ui/notification/slice.ts [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+  name: 'notification',
+  initialState,
+  reducers: {
+    addMessage: (state, { payload }) => {
+      state.notifications.push(payload)
+    },
+    removeMessage: (state, { payload }) => {
+      state.notifications = state.notifications.filter((notification) => notification.id !== payload)
+    },
+    // removeMessage: (state, { payload }) => {
```
Comment (minotogna) [inline client/store/ui/notification/slice.ts:18]: remove comments

**Review by minotogna — CHANGES_REQUESTED** [review]

**Review by minotogna — CHANGES_REQUESTED** [review]
do not use X as icon for error notifications. that confuses with the X to close the dialog. 
see evergreen for icon styles .

**Review by minotogna — CHANGES_REQUESTED** [review]
plus, add the toaster in the error Ajax middleware .

File: client/store/axiosErrorMiddleware.ts [hunk 7]
```diff
@@ -1,14 +1,21 @@
 import axios, { AxiosStatic } from 'axios'
+import { Action, Dispatch, Middleware, MiddlewareAPI } from 'redux'
+import { v4 as uuidv4 } from 'uuid'
```
Comment (minotogna) [inline client/store/axiosErrorMiddleware.ts:3]: dont we have UUIDs utils ?

**Review by minotogna — CHANGES_REQUESTED** [review]

File: client/components/Toaster/Toaster.scss [hunk 8]
```diff
@@ -0,0 +1,66 @@
+@import 'client/style/partials';
+
+.toaster-container {
+  max-width: 240px;
```
[the remark this answers is not in the document]
Comment (minotogna) [reply client/components/Toaster/Toaster.scss:4]: plus, I would use width. check 80vw or 90vw

File: client/store/axiosErrorMiddleware.ts (hunk 7 above)
Comment (rfontanarosa) [reply client/store/axiosErrorMiddleware.ts:3]: I don't think so. It seems there is no one.

Comment (rfontanarosa) [reply client/store/axiosErrorMiddleware.ts]: done.

File: client/components/Toaster/Toaster.scss (hunk 8 above)
[the remark this answers is not in the document]
Comment (rfontanarosa) [reply client/components/Toaster/Toaster.scss]: what if using 80vw with mobile and 50vw with the others?

File: client/components/Toaster/Toaster.scss [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+    }
+  }
+  &.warning {
+    color: rgb(153, 106, 19);
+    background-color: rgb(255, 250, 242);
+    border-color: rgb(255, 176, 32);
+    .toast-icon {
+      color: rgb(153, 106, 19);
+    }
+  }
```
Comment (minotogna) [inline client/components/Toaster/Toaster.scss:56]: use sass variables for colors and sizes like margins

File: server/utils/requests.ts [hunk 10]
```diff
@@ -15,14 +15,13 @@ export const sendOk = (res: any, value = {}) => res.json(value)
...
 export const sendErr = (res: any, err?: any, statusCode = 500) => {
-  console.error(err)
   if (err instanceof AccessControlException) {
     // @ts-ignore
     res.status(403).json({ error: err.error })
   } else if (typeof err === 'string') {
     res.status(statusCode).json({ error: err })
   } else {
-    res.status(statusCode).json({ error: 'Could not serve', err })
+    res.status(statusCode).json({ error: 'Could not serve' })
```
Comment (minotogna) [inline server/utils/requests.ts:24]: we lose the error here. concatenate the err

**Review by minotogna — CHANGES_REQUESTED** [review]

File: server/utils/requests.ts (hunk 10 above)
Comment (rfontanarosa) [reply server/utils/requests.ts:24]: `res.status(statusCode).json({ error: `Could not serve - ${err.message ? err.message : ''}` })`

**Review by minotogna — APPROVED** [review]

**Pull request merged**